File: go/catalogue.go
package main

import (
	"database/sql"
	"encoding/json"
	"net/http"
	"strconv"
	"time"
)

func evenements(database *sql.DB) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {

		response.Header().Set("Access-Control-Allow-Origin", "*")
		response.Header().Set("Access-Control-Allow-Headers", "Token")
		response.Header().Set("Access-Control-Allow-Methods", "GET, OPTIONS")
		if request.Method == http.MethodOptions {
			response.WriteHeader(http.StatusOK)
			return
		}

		rows, err := database.Query("SELECT id_evenement, nom, date, description, tarif FROM evenement order by date")

		if err != nil {
			http.Error(response, "Erreur lors de la selection des evenements de la base de données", http.StatusInternalServerError)
			return
		} else {
			var evenements []evenement

			for rows.Next() {
				var e evenement
				var dateSQL string
				var id int

				err := rows.Scan(&id, &e.Nom, &dateSQL, &e.Description, &e.Tarif)
				if err != nil {
					http.Error(response, "Erreur lors de la selection des evenements : "+err.Error(), http.StatusInternalServerError)
					return
				}
				t, err := time.Parse("2006-01-02 15:04:05", dateSQL)
				if err != nil {
					http.Error(response, "Erreur lors du parsing de la date : "+err.Error(), http.StatusInternalServerError)
					return
				}
				e.Date = t.Format("02/01/2006 || 15:04")

				var rej string

				auth := request.Header.Get("Token")
				userrequest, err := database.Prepare("SELECT re.id_evenement FROM reference_evenement re JOIN utilisateur u ON u.id_utilisateur = re.id_utilisateur WHERE u.token = ? AND re.id_evenement = ?")
				if err != nil {
					http.Error(response, "Erreur lors des jointures d'evenements", http.StatusInternalServerError)
					return
				}
				rowsuser := userrequest.QueryRow(auth, id)
				err = rowsuser.Scan(&rej)
				if err != nil {
					e.Rejoindre = "Rejoindre"
				} else {
					e.Rejoindre = "Quitter"
				}

				e.ID = id

				evenements = append(evenements, e)
			}
			if len(evenements) == 0 {
				json.NewEncoder(response).Encode(result{
					Message: "Aucun evenement pour le moment",
				})
				return
			}
			response.Header().Set("Content-Type", "application/json")
			json.NewEncoder(response).Encode(list{
				Evenement: evenements,
			})
		}
	}
}

func services(database *sql.DB) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {

		response.Header().Set("Access-Control-Allow-Origin", "*")
		response.Header().Set("Access-Control-Allow-Headers", "Token")
		response.Header().Set("Access-Control-Allow-Methods", "GET, OPTIONS")
		if request.Method == http.MethodOptions {
			response.WriteHeader(http.StatusOK)
			return
		}

		rows, err := database.Query("SELECT id_service, nom, description, tarif FROM service")
		if err != nil {
			http.Error(response, "Erreur lors de la selection des services de la base de données", http.StatusInternalServerError)
			return
		} else {
			var services []service

			for rows.Next() {
				var s service
				var id int

				err := rows.Scan(&id, &s.Nom, &s.Description, &s.Tarif)
				if err != nil {
					http.Error(response, "Erreur lors de la selection des services : "+err.Error(), http.StatusInternalServerError)
					return
				}
				var rej string
				var rej2 string
				auth := request.Header.Get("Token")
				userrequest, err := database.Prepare("SELECT rs.id_service FROM reference_service rs JOIN utilisateur u ON u.id_utilisateur = rs.id_utilisateur WHERE u.token = ? AND rs.id_service = ?")
				if err != nil {
					http.Error(response, "Erreur lors des jointures d'evenements", http.StatusInternalServerError)
					return
				}
				otherrequest, err := database.Prepare("SELECT id_service FROM reference_service WHERE id_service = ?")
				if err != nil {
					http.Error(response, "Erreur lors des jointures d'evenements (2)", http.StatusInternalServerError)
					return
				}
				rowsuser := userrequest.QueryRow(auth, id)
				err = rowsuser.Scan(&rej)
				rowsother := otherrequest.QueryRow(id)
				err2 := rowsother.Scan(&rej2)
				if err != nil {
					if err2 != nil {
						s.Rejoindre = "Rejoindre"
					} else {
						s.Rejoindre = "Indisponible"
					}
				} else {
					s.Rejoindre = "Quitter"
				}

				s.ID = id

				services = append(services, s)
			}
			if len(services) == 0 {
				json.NewEncoder(response).Encode(result{
					Message: "Aucun service pour le moment",
				})
				return
			}
			response.Header().Set("Content-Type", "application/json")
			json.NewEncoder(response).Encode(list{
				Service: services,
			})
		}
	}
}

func articles(database *sql.DB) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {

		response.Header().Set("Access-Control-Allow-Origin", "*")
		response.Header().Set("Access-Control-Allow-Methods", "GET, OPTIONS")
		if request.Method == http.MethodOptions {
			response.WriteHeader(http.StatusOK)
			return
		}

		rows, err := database.Query("SELECT titre, description, prix FROM article")
		if err != nil {
			http.Error(response, "Erreur lors de la selection des articles de la base de données", http.StatusInternalServerError)
			return
		} else {
			var articles []article

			for rows.Next() {
				var a article

				err := rows.Scan(&a.Nom, &a.Description, &a.Prix)
				if err != nil {
					http.Error(response, "Erreur lors de la selection des articles : "+err.Error(), http.StatusInternalServerError)
					return
				}

				articles = append(articles, a)
			}
			if len(articles) == 0 {
				json.NewEncoder(response).Encode(result{
					Message: "Aucun article pour le moment",
				})
				return
			}
			response.Header().Set("Content-Type", "application/json")
			json.NewEncoder(response).Encode(list{
				Article: articles,
			})
		}
	}
}

func evenements_patch(database *sql.DB) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {

		response.Header().Set("Access-Control-Allow-Origin", "*")
		response.Header().Set("Access-Control-Allow-Headers", "Content-Type, Token")
		response.Header().Set("Access-Control-Allow-Methods", "PATCH, OPTIONS")
		if request.Method == http.MethodOptions {
			response.WriteHeader(http.StatusOK)
			return
		}

		id, conversionError := strconv.Atoi(request.PathValue("id"))

		if conversionError != nil {
			http.Error(response, "Erreur lors de la récupération de l'identifiant de l'événement", http.StatusInternalServerError)
			return
		}

		selectstatement, selecterr := database.Prepare("SELECT id_utilisateur FROM utilisateur WHERE token = ?")
		if selecterr != nil {
			http.Error(response, "Erreur lors de la récupération des informations de l'utilisateur", http.StatusInternalServerError)
			return
		}
		var id_user int
		selectstatement.QueryRow(request.Header.Get("Token")).Scan(&id_user)
		var etat etat
		json.NewDecoder(request.Body).Decode(&etat)
		verifstatement, veriferr := database.Prepare("SELECT id_evenement FROM reference_evenement WHERE id_utilisateur = ? AND id_evenement = ?")
		if veriferr != nil {
			http.Error(response, "Erreur lors de la vérification de l'état de l'événement pour l'utilisateur", http.StatusInternalServerError)
			return
		}
		var id_verif int
		err := verifstatement.QueryRow(id_user, id).Scan(&id_verif)
		var state string

		if err != nil {
			if (err == sql.ErrNoRows) && (etat.State == "join") {
				insertstatement, inserterr := database.Prepare("INSERT INTO reference_evenement (id_utilisateur, id_evenement) VALUES (?, ?)")
				if inserterr != nil {
					http.Error(response, "Erreur lors de l'insertion de la référence de l'événement pour l'utilisateur", http.StatusInternalServerError)
					return
				}
				_, err = insertstatement.Exec(id_user, id)
				if err != nil {
					http.Error(response, "Erreur lors de l'insertion de la référence de l'événement pour l'utilisateur", http.StatusInternalServerError)
					return
				}
				state = "rejoint"
			}
		} else if etat.State == "leave" {
			deletestatement, deleteerr := database.Prepare("DELETE FROM reference_evenement WHERE id_utilisateur = ? AND id_evenement = ?")
			if deleteerr != nil {
				http.Error(response, "Erreur lors de la suppression de la référence de l'événement pour l'utilisateur", http.StatusInternalServerError)
				return
			}
			_, err = deletestatement.Exec(id_user, id)
			if err != nil {
				http.Error(response, "Erreur lors de la suppression de la référence de l'événement pour l'utilisateur", http.StatusInternalServerError)
				return
			}
			state = "quitté"
		} else {
			return
		}
		response.Header().Set("Content-Type", "application/json")
		json.NewEncoder(response).Encode(result{
			Message: "Evenement " + state + " avec succès",
		})
	}
}

func services_patch(database *sql.DB) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {

		response.Header().Set("Access-Control-Allow-Origin", "*")
		response.Header().Set("Access-Control-Allow-Headers", "Content-Type, Token")
		response.Header().Set("Access-Control-Allow-Methods", "PATCH, OPTIONS")
		if request.Method == http.MethodOptions {
			response.WriteHeader(http.StatusOK)
			return
		}

		id, conversionError := strconv.Atoi(request.PathValue("id"))

		if conversionError != nil {
			http.Error(response, "Erreur lors de la récupération de l'identifiant du service", http.StatusInternalServerError)
			return
		}

		selectstatement, selecterr := database.Prepare("SELECT id_utilisateur FROM utilisateur WHERE token = ?")
		if selecterr != nil {
			http.Error(response, "Erreur lors de la récupération des informations de l'utilisateur", http.StatusInternalServerError)
			return
		}
		var id_user int
		selectstatement.QueryRow(request.Header.Get("Token")).Scan(&id_user)
		var etat etat
		json.NewDecoder(request.Body).Decode(&etat)
		verifstatement, veriferr := database.Prepare("SELECT id_utilisateur FROM reference_service WHERE id_service = ?")
		if veriferr != nil {
			http.Error(response, "Erreur lors de la vérification de l'état du service pour l'utilisateur", http.StatusInternalServerError)
			return
		}

		var id_user_verif int
		err := verifstatement.QueryRow(id).Scan(&id_user_verif)
		var state string

		if err != nil {
			if (err == sql.ErrNoRows) && (etat.State == "join") {
				insertstatement, inserterr := database.Prepare("INSERT INTO reference_service (id_utilisateur, id_service) VALUES (?, ?)")
				if inserterr != nil {
					http.Error(response, "Erreur lors de l'insertion de la référence du service pour l'utilisateur", http.StatusInternalServerError)
					return
				}
				_, err = insertstatement.Exec(id_user, id)
				if err != nil {
					http.Error(response, "Erreur lors de l'insertion de la référence du service pour l'utilisateur", http.StatusInternalServerError)
					return
				}
				state = "rejoint"
			}
		} else if etat.State == "leave" {
			deletestatement, deleteerr := database.Prepare("DELETE FROM reference_service WHERE id_utilisateur = ? AND id_service = ?")
			if deleteerr != nil {
				http.Error(response, "Erreur lors de la suppression de la référence du service pour l'utilisateur", http.StatusInternalServerError)
				return
			}
			_, err = deletestatement.Exec(id_user, id)
			if err != nil {
				http.Error(response, "Erreur lors de la suppression de la référence du service pour l'utilisateur", http.StatusInternalServerError)
				return
			}
			state = "quitté"
		} else {
			return
		}
		response.Header().Set("Content-Type", "application/json")
		json.NewEncoder(response).Encode(result{
			Message: "Service " + state + " avec succès",
		})

	}
}

File: php/catalogue.php
<?php session_start(); include 'api_config.php'; ?>
<script src="online.js"></script>
<script>
loginUser("online", localStorage.getItem('token')); 
</script>

<script>
async function listCatalogue(token) {
    const base = (window.API_BASE || 'http://localhost:9000');

    const evenements_r = await fetch(base + "/evenements", {
        method: "GET",
        headers: {"Token": token}
    });

    if (!evenements_r.ok) {
        const html = await evenements_r.text();
        document.getElementById("error").innerHTML = "<h1>Erreur " + evenements_r.status + "</h1>" + html;
        return
    }

    const evenement_list = await evenements_r.json();
    const evenement  = document.getElementById("evenements")

    if (evenement_list.message){
        evenement.innerHTML = evenement_list.message
    } else {
        let html = "<table border = 1><tr><th>Nom de l'événement</th><th>Description</th><th>Date de l'événement</th><th></th></tr>";
        evenement_list.evenement.forEach(evenement => {
            if (evenement.rejoindre == "Rejoindre") {
                rej = "<a href='etat.php?type=evenements&state=join&id=" + evenement.id + "'>" + evenement.rejoindre + "</a>"
            } else {
                rej = "<a href='etat.php?type=evenements&state=leave&id=" + evenement.id + "'>" + evenement.rejoindre + "</a>"
            }
            html += "<tr><td>" + evenement.nom + "</td><td>" + evenement.description + "</td><td>" + evenement.date + "</td><td>" + rej + "</td></tr>"
        });
        html += "</table>";
        evenement.innerHTML = html;
    }

    const services_r = await fetch(base + "/services", {
        method: "GET",
        headers: {"Token": token}
    });

    if (!services_r.ok) {
        const html = await services_r.text();
        document.getElementById("error").innerHTML = "<h1>Erreur " + services_r.status + "</h1>" + html;
        return
    }

    const service_list = await services_r.json();
    const service  = document.getElementById("services")
    
    if (service_list.message){
        service.innerHTML = service_list.message
    } else {
        let html = "<table border = 1><tr><th>Nom du service</th><th>Description</th><th>Tarif</th><th></th></tr>";
        service_list.service.forEach(service => {
            if (service.rejoindre == "Rejoindre") {
                rej = "<a href='etat.php?type=services&state=join&id=" + service.id + "'>" + service.rejoindre + "</a>"
            } else if (service.rejoindre == "Quitter") {
                rej = "<a href='etat.php?type=services&state=leave&id=" + service.id + "'>" + service.rejoindre + "</a>"
            } else {
                rej = service.rejoindre
            }
            html += "<tr><td>" + service.nom + "</td><td>" + service.description + "</td><td>" + service.tarif + " €</td><td>" + rej + "</td></tr>"
        });
        html += "</table>";
        service.innerHTML = html;
    }

    const articles_r = await fetch(base + "/articles", {
        method: "GET",
    });

    if (!articles_r.ok) {
        const html = await articles_r.text();
        document.getElementById("error").innerHTML = "<h1>Erreur " + articles_r.status + "</h1>" + html;
        return
    }

    const article_list = await articles_r.json();
    const article  = document.getElementById("articles")

    if (article_list.message){
        article.innerHTML = article_list.message
    } else {
        let html = "<table border = 1><tr><th>Nom de l'article</th><th>Description</th><th>Prix</td></tr>";
        article_list.article.forEach(article => {
            html += "<tr><td>" + article.nom + "</td><td>" + article.description + "</td><td>" + article.prix + " €</td></tr>"
        });
        html += "</table>";
        article.innerHTML = html;
    }
}
listCatalogue(localStorage.getItem('token'));
</script>

<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Conseil</title>
</head>
<body>

<?php include 'header/header.php';

echo "<h1> Catalogue </h1>";

if (isset($_GET['message']) && isset($_SESSION['state'])) {
    echo "<h2>" . $_GET['message'] . "</h2>";
    unset($_SESSION['state']);
} ?>

<h2> Evenements </h2>
<div id = "evenements"></div>

<h2> Services </h2>
<div id = "services"></div>

<h2> Article </h2>
<div id = "articles"></div>

<div id = "error"></div>
<br>

</body>
<?php include 'footer/footer.php'; ?>
</html>
